File: stdlib/Lib/locale.py
"""Pure Python implementation of the locale module.

Interface to the locale services.
"""

import os


# Locale categories
LC_CTYPE = 0
LC_NUMERIC = 1
LC_TIME = 2
LC_COLLATE = 3
LC_MONETARY = 4
LC_MESSAGES = 5
LC_ALL = 6

# Conversion characters
CHAR_MAX = 127


_current_locale = 'C'


def setlocale(category, locale=None):
    """Set the locale for the given category."""
    global _current_locale
    if locale is None:
        return _current_locale
    if locale == '' or locale == 'C' or locale == 'POSIX':
        _current_locale = 'C'
    else:
        _current_locale = locale
    return _current_locale


def getlocale(category=LC_CTYPE):
    """Return current locale for the given category as (language, encoding)."""
    if _current_locale in ('C', 'POSIX'):
        return (None, None)
    parts = _current_locale.split('.')
    lang = parts[0] if parts else None
    enc = parts[1] if len(parts) > 1 else None
    return (lang, enc)


def getdefaultlocale(envvars=('LC_ALL', 'LC_CTYPE', 'LANG', 'LANGUAGE')):
    """Return (language, encoding) for the default locale."""
    for var in envvars:
        val = os.environ.get(var)
        if val:
            parts = val.split('.')
            lang = parts[0]
            enc = parts[1] if len(parts) > 1 else 'UTF-8'
            return (lang, enc)
    return ('en_US', 'UTF-8')


def getpreferredencoding(do_setlocale=True):
    """Return the user's preferred encoding."""
    return 'UTF-8'


def localeconv():
    """Return a dictionary of locale conventions."""
    return {
        'decimal_point': '.',
        'grouping': [],
        'thousands_sep': '',
        'currency_symbol': '',
        'int_curr_symbol': '',
        'mon_decimal_point': '',
        'mon_thousands_sep': '',
        'mon_grouping': [],
        'positive_sign': '',
        'negative_sign': '-',
        'int_frac_digits': CHAR_MAX,
        'frac_digits': CHAR_MAX,
        'p_cs_precedes': CHAR_MAX,
        'p_sep_by_space': CHAR_MAX,
        'n_cs_precedes': CHAR_MAX,
        'n_sep_by_space': CHAR_MAX,
        'p_sign_posn': CHAR_MAX,
        'n_sign_posn': CHAR_MAX,
    }


def normalize(localename):
    """Normalize locale name."""
    if not localename:
        return localename
    return localename.replace('-', '_')


def resetlocale(category=LC_ALL):
    """Reset locale to default."""
    setlocale(category, '')


def format_string(fmt, val, grouping=False, monetary=False):
    """Format a string using locale-aware formatting."""
    if isinstance(val, tuple):
        return fmt % val
    return fmt % (val,)


def currency(val, symbol=True, grouping=False, international=False):
    """Format a number as currency."""
    return '%.2f' % val


def str(val):
    """Format a number with locale decimal point."""
    return '%g' % val


def atof(string, func=float):
    """Parse a string to a float using locale settings."""
    return func(string)


def atoi(string):
    """Parse a string to an integer using locale settings."""
    return int(string)


def getencoding():
    """Get the current locale encoding."""
    return 'UTF-8'

File: crates/ferrython-stdlib/src/sys_modules/locale.rs
use compact_str::CompactString;
use ferrython_core::error::PyException;
use ferrython_core::object::{
    make_builtin, make_module, PyCell, PyObject, PyObjectMethods, PyObjectRef,
};
use indexmap::IndexMap;
use std::rc::Rc;

static CURRENT_CTYPE_LOCALE: std::sync::LazyLock<parking_lot::RwLock<(String, String)>> =
    std::sync::LazyLock::new(|| {
        parking_lot::RwLock::new(("C".to_string(), "ANSI_X3.4-1968".to_string()))
    });

pub fn get_current_ctype_locale() -> (String, String) {
    CURRENT_CTYPE_LOCALE.read().clone()
}

// ── locale module ──

pub fn create_locale_module() -> PyObjectRef {
    // Detect system locale from environment
    fn get_system_locale() -> (String, String) {
        let lang = std::env::var("LC_ALL")
            .or_else(|_| std::env::var("LC_CTYPE"))
            .or_else(|_| std::env::var("LANG"))
            .unwrap_or_else(|_| "C".to_string());
        if lang == "C" || lang == "POSIX" || lang.is_empty() {
            return ("C".to_string(), "ANSI_X3.4-1968".to_string());
        }
        // Parse "en_US.UTF-8" into ("en_US", "UTF-8")
        if let Some(dot) = lang.find('.') {
            let locale_name = &lang[..dot];
            let encoding =
                lang[dot + 1..].trim_end_matches(|c: char| !c.is_alphanumeric() && c != '-');
            (locale_name.to_string(), encoding.to_string())
        } else {
            (lang.clone(), "UTF-8".to_string())
        }
    }

    let current_locale: Rc<PyCell<(String, String)>> = Rc::new(PyCell::new(get_system_locale()));
    *CURRENT_CTYPE_LOCALE.write() = current_locale.read().clone();

    let cl1 = current_locale.clone();
    let getlocale_fn = PyObject::native_closure("getlocale", move |_: &[PyObjectRef]| {
        let (name, enc) = cl1.read().clone();
        Ok(PyObject::tuple(vec![
            PyObject::str_val(CompactString::from(name)),
            PyObject::str_val(CompactString::from(enc)),
        ]))
    });

    let cl2 = current_locale.clone();
    let setlocale_fn = PyObject::native_closure("setlocale", move |args: &[PyObjectRef]| {
        let _category = if !args.is_empty() {
            args[0].to_int().unwrap_or(0)
        } else {
            0
        };
        if args.len() >= 2 {
            let locale_str = args[1].py_to_string();
            if locale_str.is_empty() || locale_str == "C" || locale_str == "POSIX" {
                *cl2.write() = ("C".to_string(), "ANSI_X3.4-1968".to_string());
            } else if let Some(dot) = locale_str.find('.') {
                *cl2.write() = (
                    locale_str[..dot].to_string(),
                    locale_str[dot + 1..].to_string(),
                );
            } else {
                *cl2.write() = (locale_str.clone(), "UTF-8".to_string());
            }
            *CURRENT_CTYPE_LOCALE.write() = cl2.read().clone();
        }
        let (name, enc) = cl2.read().clone();
        Ok(PyObject::str_val(CompactString::from(format!(
            "{}.{}",
            name, enc
        ))))
    });

    let cl3 = current_locale.clone();
    let localeconv_fn = PyObject::native_closure("localeconv", move |_: &[PyObjectRef]| {
        let (name, _) = cl3.read().clone();
        let is_c = name == "C" || name == "POSIX";
        let mut conv = IndexMap::new();
        conv.insert(
            CompactString::from("decimal_point"),
            PyObject::str_val(CompactString::from(".")),
        );
        conv.insert(
            CompactString::from("thousands_sep"),
            PyObject::str_val(CompactString::from(if is_c { "" } else { "," })),
        );
        conv.insert(
            CompactString::from("grouping"),
            PyObject::list(if is_c {
                vec![]
            } else {
                vec![PyObject::int(3), PyObject::int(0)]
            }),
        );
        conv.insert(
            CompactString::from("int_curr_symbol"),
            PyObject::str_val(CompactString::from("")),
        );
        conv.insert(
            CompactString::from("currency_symbol"),
            PyObject::str_val(CompactString::from("")),
        );
        conv.insert(
            CompactString::from("mon_decimal_point"),
            PyObject::str_val(CompactString::from(".")),
        );
        conv.insert(
            CompactString::from("mon_thousands_sep"),
            PyObject::str_val(CompactString::from(if is_c { "" } else { "," })),
        );
        conv.insert(CompactString::from("p_sign_posn"), PyObject::int(1));
        conv.insert(CompactString::from("n_sign_posn"), PyObject::int(1));
        let dict = PyObject::dict_from_pairs(
            conv.into_iter()
                .map(|(k, v)| (PyObject::str_val(k), v))
                .collect(),
        );
        Ok(dict)
    });

    let normalize_fn = make_builtin(|args: &[PyObjectRef]| {
        if args.is_empty() {
            return Err(PyException::type_error("normalize() requires 1 argument"));
        }
        let s = args[0].py_to_string();
        // Simple normalization: "en_US" → "en_US.UTF-8"
        if !s.contains('.') {
            Ok(PyObject::str_val(CompactString::from(format!(
                "{}.UTF-8",
                s
            ))))
        } else {
            Ok(PyObject::str_val(CompactString::from(s)))
        }
    });

    make_module(
        "locale",
        vec![
            ("getlocale", getlocale_fn),
            ("setlocale", setlocale_fn),
            ("localeconv", localeconv_fn),
            ("normalize", normalize_fn),
            (
                "getpreferredencoding",
                make_builtin(|_| Ok(PyObject::str_val(CompactString::from("UTF-8")))),
            ),
            (
                "getdefaultlocale",
                make_builtin(|_| {
                    let lang = std::env::var("LANG").unwrap_or_else(|_| "en_US.UTF-8".to_string());
                    let (name, enc) = if let Some(dot) = lang.find('.') {
                        (lang[..dot].to_string(), lang[dot + 1..].to_string())
                    } else {
                        (lang, "UTF-8".to_string())
                    };
                    Ok(PyObject::tuple(vec![
                        PyObject::str_val(CompactString::from(name)),
                        PyObject::str_val(CompactString::from(enc)),
                    ]))
                }),
            ),
            ("LC_CTYPE", PyObject::int(0)),
            ("LC_NUMERIC", PyObject::int(1)),
            ("LC_TIME", PyObject::int(2)),
            ("LC_COLLATE", PyObject::int(3)),
            ("LC_MONETARY", PyObject::int(4)),
            ("LC_MESSAGES", PyObject::int(5)),
            ("LC_ALL", PyObject::int(6)),
            ("CHAR_MAX", PyObject::int(127)),
            (
                "Error",
                PyObject::builtin_type(CompactString::from("locale.Error")),
            ),
            (
                "str",
                make_builtin(|args: &[PyObjectRef]| {
                    if args.is_empty() {
                        return Ok(PyObject::str_val(CompactString::from("")));
                    }
                    Ok(PyObject::str_val(CompactString::from(
                        args[0].py_to_string(),
                    )))
                }),
            ),
            (
                "atof",
                make_builtin(|args: &[PyObjectRef]| {
                    if args.is_empty() {
                        return Err(PyException::type_error("atof() requires 1 argument"));
                    }
                    let s = args[0].py_to_string().replace(',', "");
                    let f: f64 = s.parse().map_err(|_| {
                        PyException::value_error(format!("could not convert '{}' to float", s))
                    })?;
                    Ok(PyObject::float(f))
                }),
            ),
            (
                "atoi",
                make_builtin(|args: &[PyObjectRef]| {
                    if args.is_empty() {
                        return Err(PyException::type_error("atoi() requires 1 argument"));
                    }
                    let s = args[0].py_to_string().replace(',', "");
                    let n: i64 = s.parse().map_err(|_| {
                        PyException::value_error(format!("could not convert '{}' to int", s))
                    })?;
                    Ok(PyObject::int(n))
                }),
            ),
        ],
    )
}

// ── inspect module (stub) ──

// ── getpass module ───────────────────────────────────────────────────
